Support `@test_throws` just checking the error message

timholy opened the pull request:
With this PR,
```julia
@test_throws "reducing over an empty collection" reduce(+, ())
```
allows you to check the displayed error message without caring about
the details of which specific Exception subtype is used.

Now let me give a little more background: as mentioned in #41885, the exception type of errors has changed repeatedly. IMO, the exception type's main purpose is to ensure that the correct message gets delivered to the user. The fact that `@test_throws` focused on the exception *type* is therefore basically a mis-feature, because it should really be testing the thing we actually care about rather than the internals of the mechanism used to achieve it. This PR allows one to focus on what actually matters.

@vtjnash, I normally wouldn't ask you to review something so simple as this, but in light of the scary bugs you and Jeff were squashing in #37809 and #38127, can I ask you to take a peek and make sure I haven't broken anything?

timholy committed:
Support `@test_throws` just checking the error message

With this PR,

    @test_throws "reducing over an empty collection" reduce(+, ())

allows you to check the displayed error message without caring about
the details of which specific Exception subtype is used.

The pattern-matching options are the same as for `@test_warn`.

timholy commented on NEWS.md:
```diff
@@ -37,6 +37,9 @@ New library functions
 
+* `@test_throws "some message" triggers_error()` can now be used to check whether the displayed error text
+  contains "some message" regardless of the specific exception type.
+  Regular expressions are also supported. ([#?????])
```
```suggestion
  Regular expressions are also supported. ([#41888)
```

timholy commented on stdlib/Test/src/Test.jl:
```diff
@@ -671,6 +687,11 @@ julia> @test_throws DimensionMismatch [1, 2, 3] + [1, 2]
+julia> @test_throws "Try sqrt(Complex" sqrt(-1)
+Test Passed
+  Expression: sqrt(-1)
+     Message: "DomainError with -1.0:\nsqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x))."
```
```suggestion
     Message: "DomainError with -1.0:\\nsqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x))."
```

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -704,7 +725,16 @@ function do_test_throws(result::ExecutionResult, orig_expr, extype)
-        if isa(extype, Type)
+        if isa(extype, AbstractString) || isa(extype, AbstractPattern)
+            exc = '"' * escape_string(sprint(showerror, exc)) * '"'
+            if isa(extype, AbstractString)
```
Can we use the test_warn matchers here for consistency?

timholy replied:
Done

vtjnash commented:
You might need to also edit the serialize methods to preserve this type too. Or use an extra bool field instead of branching on type?

tkf commented:
Are these (arguably very wacky) cases still supported?

```julia
julia> @test_throws "\"" throw("\"")
Test Passed
      Thrown: String

julia> @test_throws Returns(false) throw(Returns(false))
Test Passed
  Expression: throw(Returns(false))
      Thrown: Returns{Bool}
```

Not sure if they should be working, but maybe good edge cases to test against?

(The functionality added by this PR looks very useful BTW :+1:)

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -697,13 +717,17 @@ function do_test_throws(result::ExecutionResult, orig_expr, extype)
+        end
         if isa(extype, Type)
             success =
                 if from_macroexpand && extype == LoadError && exc isa Exception
```
won't the code movement break this depwarn?

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -712,24 +736,30 @@ function do_test_throws(result::ExecutionResult, orig_expr, extype)
+                    break
                 end
             end
+        elseif isa(extype, Exception)
+        else
```
```suggestion
        elseif !isa(extype, Exception)
```
(nit: looks odd otherwise, like something was missing)

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -712,24 +736,30 @@ function do_test_throws(result::ExecutionResult, orig_expr, extype)
+            message_only = true
+            exc = sprint(showerror, exc)
+            success = contains_warn(exc, extype)
+            exc = '"' * escape_string(exc) * '"'
```
```suggestion
            exc = repr(exc)
```

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -712,24 +736,30 @@ function do_test_throws(result::ExecutionResult, orig_expr, extype)
+            success = contains_warn(exc, extype)
+            exc = '"' * escape_string(exc) * '"'
+            if isa(extype, AbstractString)
+                extype = '"' * escape_string(extype) * '"'
```
```suggestion
                extype = repr(extype)
```

vtjnash reviewed (commented):
LGTM!

timholy commented:
> Are these (arguably very wacky) cases still supported?

```julia
julia> @test_throws "\"" throw("\"")
Test Passed
  Expression: throw("\"")
     Message: "\"\\\"\""

julia> @test_throws Returns(false) throw(Returns(false))
Test Failed at REPL[6]:1
  Expression: throw(Returns(false))
    Expected: < match function >
     Message: "Returns{Bool}(false)"
ERROR: There was an error during testing
```

So the first yes, the second no. But is the second actually desirable?

```julia
julia> throw(()->startswith("cat", "nope"))
ERROR: var"#11#12"()
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```
and since most people will probably use the functional form anonymously, I think we'd get lots of opaque error messages.

To help me wrap my head around this, can you give me an example of a use of throwing something besides an `Exception`?

timholy committed:
Address review comments

Co-authored by: Jameson Nash <vtjnash@gmail.com>
Co-authored by: Takafumi Arakaki <aka.tkf@gmail.com>

timholy commented:
Nvm, it was pretty easy to get both of those to pass: I made this logic conditional on the thrown object being an `Exception` but the match object not being an exception or exception type.

timholy committed:
Better NEWS and docstring

timholy commented:
What's up with all the buildkite failures? Is this something I broke? I'm seeing similar failures in some other jobs that ran recently.

ericphanson commented:
Would be nice to have a compat annotation for this feature in the `@test_throws ` docstring
